Add  Alexa.ChannelController functions for media players

Dilbert66 opened the pull request:
Added missing Alexa.ChannelController functions. Specifically Change Channel and SkipChannel commands. These functions will call the play_media function in a media_player app if it has the capability published and pass on the  channel# or channel name. The selected media player can then use this to select the channel on the device it is associated to.

Modified the existing Alexa.StepSpeaker Setvolume function to actually do a stepped volume change using the steps sent by Alexa. The Alexa default step of 10 for a simple volume up/down can be changed via an exposed media_player attribute called volume_step_default.  The default is set to 1.
Any other value than default will be sent  as a looped sequence of volume up/down calls to the media_player.

Modifed the test script test_smart_home.py to add the missing payload attribute volumeStepDefault and test for it.  Also added ChannelController tests.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]



If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

Checks the Channel controller box in  #24579

Dilbert66 committed:
Added missing Alexa.ChannelController functions. Specifically ChangeChannel
and SkipChannel commands. These functions will call the play_media function
 in a media_player app if it has the capability published and pass on the
  channel# or channel name. The selected media player can then use this to
   select the channel on the device it is associated to.
Modified the existing Alexa.StepSpeaker Setvolume function to actually do
a stepped volume change using the steps sent by Alexa. The Alexa default
 step of 10 for a simple volume up/down can be changed via an exposed
 media_player attribute called volume_step_default.
 The default is set to 1. Any other value then default will be sent
 as sequential volume up /down to the media_player.

Dilbert66 commented:
This is a a resubmit of my previous pull request. Due to issues with my dev environment , I felt it best to start fresh.  I now have the tox  , flake8, pylint programs running correctly.

Dilbert66 committed:
The test code has some weird behaviour with passed boolean values.  Had to surround them in quotes for the tests to pass properly.

Dilbert66 committed:
Reverted test_smart_home.py change. Issue was not the boolean value but the behavior in the handler. The test suite does not like multiple await calls in a loop. Will investigate further.  The handler code works though.

Dilbert66 committed:
Added ChannelController/SkipChannels test in test_smart_home.py
Added test for callSign payload attribute.

Dilbert66 committed:
Modified smart home test to allow more than one call to services

Dilbert66 committed:
Added more tests for ChannelChange functions for various payload options.
Removed name options from metadata payload section. not needed.

Dilbert66 committed:
Reverted assert call change in alexa test __init__.py back to ==1. Not sure if it was the cause of the pytest's failing on github

Dilbert66 commented:
I'm totally confused by the check system???  Trying to figure why it fails is an incredible frustration?  Mosts tests passed before. I do a simple commit and then i start getting "docker pull errors" for all checks ?? What does that mean?  The code changes pass lint/flake8, etc.

Dilbert66 committed:
Corrected a comment.  First commit after a rebase.

Dilbert66 commented:
From the logs on Azure looks like the issue is there and not with my code.

Dilbert66 committed:
Comment line change.  Also wanted to force a code check on github.

tulindo commented:
Hi @Dilbert66 I've seen your code and doesn't look bad... 
Just a comment: You implemented the volumestep and channelskip functions using a loop. What about adding a sleep between every volumeup/nexttrack command for safety?

Dilbert66 commented:
A safety delay is a good idea actually.  There is  processing delay induced with the call but I agree, i will add it.

Dilbert66 committed:
Added a loop delay in StepSpeaker and SkipChannel functions for safety

balloob commented on homeassistant/components/alexa/handlers.py:
```diff
@@ -520,21 +521,49 @@
+                    blocking=False,
+                    context=context,
+                )
+            await asyncio.sleep(sleep_delay)
```
We shouldn't sleep. It's up to the receiving integration to throttle when necessary.

balloob commented on tests/components/alexa/test_smart_home.py:
```diff
@@ -425,8 +425,9 @@ def assert_endpoint_capabilities(endpoint, *interfaces):
         {
             "friendly_name": "Test media player",
-            "supported_features": 0x59BD,
+            "supported_features": 0x59BD | 16384 | 4096 | 512 | 32 | 16 | 1,
```
Please use the constants instead of the values of them.

balloob commented on homeassistant/components/alexa/handlers.py:
```diff
@@ -520,21 +521,49 @@
+
+    if "volume_steps_default" in entity.attributes:
+        try:
+            default_steps = int(entity.attributes["volume_steps_default"])
```
We should not introduce new attributes to an integration for specific use cases. We like to put a little more thought into the attributes that we support. Please remove this.

You can open an architecture issue if you want to propose a new attribute for an entity. I don't think this warrants the addition of a newly supported attribute though.

Dilbert66 replied:
It's just a nice to have but definitively not a required one as all one has to say is "Alexa, turn up the volume by 5 on tv" instead of just 'Alexa, turn up the volume in tv".   
There are a few additional things that can be added to the media_player as suggested by others for more specific control of various hardware but as you noted in this case, all functionality can be accomplished by existing attributes.  I've remove it and kept it at basics. Thanks.

balloob commented on homeassistant/components/alexa/handlers.py:
```diff
@@ -928,3 +956,92 @@ def temperature_from_object(hass, temp_obj, interval=False):
+        if channel > 0:
+            await hass.services.async_call(
+                entity.domain,
+                SERVICE_MEDIA_NEXT_TRACK,
```
Define the service name in a variable prior to executing the range instead of defining two code blocks.

balloob commented on homeassistant/components/alexa/handlers.py:
```diff
@@ -520,21 +521,49 @@
+            elif volume_int < 0:
+                await hass.services.async_call(
+                    entity.domain,
+                    SERVICE_VOLUME_DOWN,
```
Same, define prior to for-loop.

Dilbert66 committed:
Removed uneeded sleep from for loops.  Let remote handle delays
Moved service type decision out of for loops in ChannelController and StepSpeaker
Used constants instead of numeric values for support options in test module

Dilbert66 committed:
Change media_player const import to be more specific in source

Dilbert66 commented:
Thank you. I've made the recommended changes , re-tested and commited.

Dilbert66 committed:
Modifed test_smart_home to use media_play constants instead of hardcode valu

Dilbert66 committed:
Removed unecessary test volume_step_default attribute from test_smart_home

Dilbert66 committed:
Removed uneeded comment in StepSpeaker function.
Re-ordered constants in test_smart_home.py

Dilbert66 committed:
Modified call to media_player play_media service to use media_player constant instead of hard coded value.

Dilbert66 committed:
Changed constant use to be consistant with rest of function.

balloob reviewed (approved):
Excellent! I had just merged another PR for Alexa so there are some merge conflicts.
This PR can be merged when the conflicts have been resolved.

Great first contribution! 🐬  Welcome to the Home Assistant community 🎉

Dilbert66 committed:
Correct merge conflicts in handlers.py and capablities.py

Dilbert66 commented:
Thank you! Conflicts resolved and re-pushed.   I see that the Python37 checks are failing but from the looks of the azure logs, this looks to be an issue outside of this pull request.

balloob commented:
🎉
